documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/LoadBalancer.java

```diff
@@ -68,24 +77,24 @@ public Node getRecipient(List<Mirror.Entry> choices) {
 
         double weightSum = 0.0;
         Node selectedNode = null;
-        double position = safePosition.get();
-        for (Mirror.Entry entry : choices) {
-            NodeMetrics nodeMetrics = getNodeMetrics(entry);
+        synchronized (this) {
+            for (Mirror.Entry entry : choices) {
+                NodeMetrics nodeMetrics = getNodeMetrics(entry);
 
-            weightSum += nodeMetrics.weight;
+                weightSum += nodeMetrics.weight;
 
-            if (weightSum > position) {
-                selectedNode = new Node(entry, nodeMetrics);
-                break;
+                if (weightSum > position) {
+                    selectedNode = new Node(entry, nodeMetrics);
```

Comment:
The returned `NodeMetrics` object is shared and thus may be read outside the lock whilst concurrently being updated by another thread. Is this a problem? Should we deep-copy the metrics to make sure they're internally consistent? This is not something introduced by this PR, but just a general observation.

Reply:
That was why I made it package private. I checked all other usage. That was only in tests.